mycpp/gc_dict.h

```diff
@@ -110,42 +98,43 @@ class Dict {
   void reserve(int n);
 
   // d[key] in Python: raises KeyError if not found
-  V index_(K key);
+  V index_(K key) const;
 
   // Get a key.
   // Returns nullptr if not found (Can't use this for non-pointer types?)
-  V get(K key);
+  V get(K key) const;
 
   // Get a key, but return a default if not found.
   // expr_parse.py uses this with OTHER_BALANCE
-  V get(K key, V default_val);
+  V get(K key, V default_val) const;
 
   // Implements d[k] = v.  May resize the dictionary.
   void set(K key, V val);
 
   void update(List<Tuple2<K, V>*>* kvs);
 
-  List<K>* keys();
+  List<K>* keys() const;
 
   // For AssocArray transformations
-  List<V>* values();
+  List<V>* values() const;
 
   void clear();
 
-  // Returns the position in the array.  Used by dict_contains(), index(),
-  // get(), and set().
+  // Returns an offset into the index for given key. If the key is not already
+  // in the table and there is room, the offset to an empty slot will be
+  // returned. The caller is responsible for checking if the index slot is empty
+  // before using it.
   //
-  // For now this does a linear search.
-  // TODO:
-  // - hash functions, and linear probing.
-  // - resizing based on load factor
-  //   - which requires rehashing (re-insert all items)
-  // - Special case to intern Str* when it's hashed?  How?
-  //   - Should we have wrappers like:
-  //   - V GetAndIntern<V>(D, &string_key)
-  //   - SetAndIntern<V>(D, &string_key, value)
-  //   This will enable duplicate copies of the string to be garbage collected
-  int position_of_key(K key);
+  // Returns -1 if the dictionary is full. The caller can use this as a cue to
+  // grow the table.
+  //
+  // Used by dict_contains(), index(), get(), and set().
+  int find_key_in_index(K key) const;
```

Comment:
How about names like

- bucket_of(K key) -- well I guess it's not a bucket, since we're doing linear probing

- hash_and_probe(K key)
- probe(K key)

...

- index_of(K key) -- the `index_` member holds the index into the keys and values
- kv_index(K key)

Reply:
how about `hash_and_probe(K)` and `lookup_kv(K)`?

Reply:
- I think `hash_and_probe()` or just `probe()` are good
  - or maybe `probe_for_bucket()`
- `lookup_kv()` is OK, I think I would prefer naming after something it returns, like `find_kv_index()`

(Although also our names are a bit inconsistent, maybe it should be ProbeForBucket() for non-user facing methods ... it has to be `.get()` for Python compat, but these don't have that constraint)

If we want the most descriptive names (which we don't always, but here it seems worth distinguishing), I think Verb-Noun is good

- probe is the verb, bucket is the noun
- find is the verb, kv_index is the noun

Reply:
Let's go with `hash_and_probe()` and `find_kv_index()`. We don't really use the term bucket anywhere else in `Dict` right now so I think we should avoid introducing it, though I agree it feels natural.